Test1/src/leetcode/LeetCode22.java

```diff
@@ -0,0 +1,54 @@
+package leetcode;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * https://leetcode.com/problems/generate-parentheses/
+ * @author minibig
+ */
+public class LeetCode22 {
+	
+	private Map<Integer, List<String>> map = new HashMap<Integer, List<String>>();
+	
+	public List<String> generateParenthesis(int n) {
+		
+		List<String> strList = new ArrayList<String>();
+		strList.add("");
+		
+		return generate(strList, n);
+	}
+	
+	public List<String> generate(List<String> strList, int n) {
+		
+		if(map.get(n) != null) {
+			return map.get(n);
+		}
+		
+		List<String> list = new ArrayList<String>();
+		
+		if(n == 1) {
+			list.add("()");
+			map.put(1, list);
+			return list;
+		}else {
+			for(int i = 1 ; i < n; i++) {
+				for(String str1 : generate(strList, n-i)) {
+					if(i == 1) {
+						list.add("(" + str1 + ")");
+					}
+					for(String str2 : generate(strList, i)) {
+						list.add(str1 + str2);
+					}
+				}
+			}
+		}
+		
+		map.put(n, list.stream().distinct().collect(Collectors.toList()));
```

Comment:
The `distinct()` here is very risky because it indicates that you put duplicate answers into the list.

When you have duplicate answers, you're very likely doing something wrongly (or super inefficiently).
You should think about a solution which doesn't generate duplicate answers.
If the duplicate answers are inevitable, make sure the number of duplicate answers is controllable. 

I'm not pretty sure about this problem, but I guess there are some solutions that completely avoid duplicate answers.

Reply:
```
class Solution {
public:
    vector<string> ans;
    
    void dfs(string &s, int c1, int c2) {
        if(c1 + c2 == 0) {
            ans.push_back(s);
        } else {
            if(c1 > 0) {
                s += '(';
                dfs(s, c1-1, c2+1);
                s.pop_back();
            }
            if(c2 > 0) {
                s += ')';
                dfs(s, c1, c2-1);
                s.pop_back();
            }
        }
    }
    
    vector<string> generateParenthesis(int n) {
        string tmp = "";
        dfs(tmp, n, 0);
        return ans;
    }
};
```

This is the code which I wrote several years ago when I worked on this problem.

Reply:
I just wrote the code based on your solution.

```
class Solution {
private:
    unordered_map<int, vector<string>> answers;
    unordered_map<int, vector<string>> single_segment;
public:
    Solution() {
        // initialization
        answers[1].push_back("()");
        single_segment[1] = answers[1];
    }
    vector<string> generateParenthesis(int n) {
        if(answers.find(n) != answers.end()) {
            return answers[n];
        }
        
        vector<string> current_answers;
        for(auto answer: generateParenthesis(n-1)) {
            current_answers.push_back("(" + answer + ")");
        }
        single_segment[n] = current_answers;
        
        for(int i = 1; i < n; i++) {
            for(auto answer_first: generateParenthesis(n-i)) {
                for(auto answer_second: single_segment[i]) {
                    current_answers.push_back(answer_first + answer_second);
                }
            }
        }
        
        return answers[n] = current_answers;
    }
};
```

I think in this solution we don't need `distinct()` function call. This was written in C++.